[PUBDEV-5417] Enhancement to gcs-persistence to support Flows directory placed on GCS.

pcejrowski opened the pull request:
This PR contains of two parts:
* enhancements to GCS persist which allow to save Flow using `gs://` prefixed paths on GCS
* exposing endpoints which allow to load Flows from any persistent storage path utilizing existing `-flow_dir` parameter.

michalkurka commented on h2o-core/src/main/java/water/api/FlowLoadServlet.java:
```diff
@@ -0,0 +1,79 @@
+public class FlowLoadServlet extends HttpServlet {
+
+  @Override
+  protected void doGet(HttpServletRequest request, HttpServletResponse response) {
```
@pcejrowski Hmm, this doesn't actually handle any binary input/output. Seems to me it could be implemented as a Handler with much less code. Am I missing something?

michalkurka commented on h2o-core/src/main/java/water/api/FlowLoadServlet.java:
```diff
@@ -0,0 +1,79 @@
+      if (is == null) {
+        return;
+      }
+      H2O.getNPS().put(categoryName, keyName, is);
```
IMHO belongs somewhere else. 

This just copies data from a storage to the NodePersistantStorage. I don't think it warrants a servlet. Servlets are more for binary data that cannot be handled elsewhere.

pcejrowski commented:
@michalkurka Based on your comments in [h2oai/h2o-flow#65](https://github.com/h2oai/h2o-flow/pull/65) I removed the Servlet and left only necessary enhancements to gcs-persistence module. 

Please review.

Pscheidl commented on h2o-core/src/main/java/water/persist/PersistManager.java:
```diff
@@ -465,6 +476,8 @@ public boolean mkdirs(String path) {
       boolean b = I[Value.HDFS].mkdirs(path);
       return b;
+    } else if (isGcsPath(path)){
+      return true; // GCS doesn't support directories, so we don't need them to exist
```
This behavior should be tested. There are some notes about [subdirectories in GCS](https://cloud.google.com/storage/docs/gsutil/addlhelp/HowSubdirectoriesWork) to be found. Are we sure we do not support those ?

What if the dir is not created and we return true ? Shouldn't we return false, or even UnsupportedOperationException ?

pcejrowski replied:
The entire directory concept doesn't exist in GCS. 

> To the service, the object gs://your-bucket/abc/def.txt is just an object that happens to have "/" characters in its name.

My reasoning behind this `return true;` here is that we don't need to have directory created the way we do need in other file systems and throwing exception / returning false requires us to treat GCS exceptionally in every code which uses `PersistManager`

pcejrowski replied:
Rethought it!
We should have something for bucket creation in this place, since the path might point at the bucket. 
Will take care later.

pcejrowski replied:
Implemented `mkdirs`, so that bucket is created if doesn't exist. Directory within the bucket is ignored, since this concept is not supported by GCS.

Pscheidl commented on h2o-core/src/main/java/water/util/FileUtils.java:
```diff
@@ -31,6 +31,7 @@ public static void copyStream(InputStream is, OutputStream os, final int buffer_
           break;
         os.write(bytes, 0, count);
       }
+      os.close();
```
Note to self: Investigate this line. Why should we close a stream that was given to the method ? Is that really part of method's contract ?

pcejrowski replied:
I think that it should be here, since the entire stream is read and without this line it's not flushed to disk.

michalkurka replied:
hmm, in my opinion, it should really be a responsibility of the caller. Only the caller knows that it will not use eg. mark() and reset() methods on the stream and attempt to rewind it back.

`IOUtils.copy` is not closing the input stream. We should take it as a reference. 

Problem is there are incorrect uses of the copyStream in our code bases and streams are not closed. That should be fixed

pcejrowski replied:
Fixed in 838b5b8.

Pscheidl commented on h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java:
```diff
@@ -225,4 +227,90 @@ public void close() throws IOException {
+
+  @Override
+  public boolean delete(String path) {
+    final BlobId blob = GcsBlob.of(path).getBlobId();
```
What about cache ? Should we modify it too ? Or force to refresh it ?

michalkurka replied:
I suggest not to modify the cache. The cache is okay for "type ahead" kind of operation but it should not be used anywhere else in the code. If necessary it should be invalidated here.

pcejrowski replied:
Invalidation has been added.

Pscheidl commented on h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java:
```diff
@@ -225,4 +227,90 @@ public void close() throws IOException {
+
+  @Override
+  public boolean rename(String fromPath, String toPath) {
+    final BlobId fromBlob = GcsBlob.of(fromPath).getBlobId();
```
Cache update ?

Pscheidl commented:
Added some points to discuss. The PR looks good in general. Just a few questions to consider, including the cache management.

pcejrowski commented:
@Pscheidl Hey!
Added cache invalidations that you asked for (didn't want to play with cache updates :) ) and refactored `exist()` method, so that the the flow is not managed by exceptions.

michalkurka commented on h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java:
```diff
@@ -225,4 +227,92 @@ public void close() throws IOException {
+   */
+  @Override
+  public PersistEntry[] list(String path) {
+    List<String> matches = calcTypeaheadMatches(path, Integer.MAX_VALUE);
```
I would avoid using the typeahead cache here. Also not that current implementation of `calcTypeaheadMatches` doesnt't actually use the `limit` parameter

pcejrowski replied:
Good points!
Resolved both.
See 3ac6e31 and bcb35bf.

pcejrowski commented:
I fixed all the issues that you've brought out. 
Is that ready for merge?

Pscheidl commented on h2o-core/src/main/java/water/api/DatasetServlet.java:
```diff
@@ -47,8 +49,20 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+        os = response.getOutputStream();
+        FileUtils.copyStream(is, os, 2048);
+      } finally {
+        if (os != null) {
```
try-with block could be used, nothing important

Pscheidl commented on h2o-core/src/main/java/water/api/NpsBinServlet.java:
```diff
@@ -42,8 +44,20 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) {
+        os = response.getOutputStream();
+        FileUtils.copyStream(is, os, 2048);
+      } finally {
+        if (os != null) {
```
try-with-resources block could be used (not that important)

Pscheidl commented on h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java:
```diff
@@ -112,19 +114,25 @@ public void cleanUp() {
     final String input = GcsBlob.removePrefix(filter);
     final String[] bk = input.split("/", 2);
-    List<String> results = new ArrayList<>();
+    List<String> results = limit > 0 ? new ArrayList<String>(limit) : new ArrayList<String>();
```
Setting it to limit could allocate too many resources in the backing array in vain in some cases. Not sure if we want it. However, it makes sense to do it. On the other hand, allocating an empty ArrayList with zero-length backing array will probably result in immediate array enlargment in vain I believe. This component is not performance-critical, but I'd think set the initial capacity to 10 for example, or some arbitrary number > 2.

michalkurka replied:
When you use the default non-parametric constructor it doesn't allocate anything (it reuses a static 
"empty" array). The initialization of the backing array will actually happen after you add the first element to the array.

I think this code is fine as it is

Pscheidl commented on h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java:
```diff
@@ -225,4 +233,121 @@ public void close() throws IOException {
+    return new OutputStream() {
+      @Override
+      public void write(int b) throws IOException {
+        ByteBuffer buffer = ByteBuffer.wrap(new byte[]{(byte) b});
```
Are we sure there will always be byte ? 

Also, I was looking for other ways to write a single byte and not to wrap it with a new object each time a byte needs to be written - to no avail.

pcejrowski replied:
It might be difficult because Google Cloud Storage library accepts only `ByteBuffer`.

Pscheidl commented:
I had some thoughts in the comments, but in general I am OK with this. Let's wait for @michalkurka 's approval.

Pscheidl commented:
I had some thoughts in the comments, but in general I am OK with this. Let's wait for @michalkurka 's approval.

michalkurka commented on h2o-core/src/main/java/water/init/NodePersistentStorage.java:
```diff
@@ -123,7 +123,7 @@ public void put(String categoryName, String keyName, InputStream is) {
       }
     }
-    if (! pm.exists(tmpd)) {
+    if (! H2O.getPM().isGcsPath(tmpd) && ! pm.exists(tmpd)) {
```
I wouldn't introduce this special case handling for GCS. Frankly, I would rather remove this check completely. I don't see the purpose of checking for the existence of the directory right after we explicitly created it.

Suggestion: remove lines 126-128. Your thoughts @Pscheidl ?

Similar on lines 138-140

michalkurka commented on h2o-core/src/main/java/water/persist/PersistManager.java:
```diff
@@ -429,6 +436,8 @@ public long length(String path) {
       validateHdfsConfigured();
       long l = I[Value.HDFS].length(path);
       return l;
+    } else if (isGcsPath(path)) {
```
I think these if-statements are good for now but we should eventually refactor the code and make adding new persist backends pluggable.

pcejrowski replied:
Totally agree. Didn't want to struggle with that in this PR.

michalkurka commented:
LGTM, just minor comments

pcejrowski commented:
@michalkurka @Pscheidl hey! is there anything which stops you from merging?

michalkurka commented:
@pcejrowski thanks! LGTM

michalkurka commented:
We need to wait with merging to get the tests results.

Pscheidl commented:
Testing results will be available today. I did some manula testing and it looks good. Let's wait for the detailed ones and merge today eventually.

Pscheidl commented:
I simple NullPointer on `gs://gcp-public-data-nexrad-l2/2018/01/01/` trying to be imported in flow. 

```plain
04-16 13:49:55.009 10.0.0.9:54321        12199  #31937-23 ERRR: ; Stacktrace: [java.lang.NullPointerException,     water.persist.PersistManager.calcTypeaheadMatches(PersistManager.java:251),     water.api.TypeaheadHandler.files(TypeaheadHandler.java:12),     sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source),     sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43),     java.lang.reflect.Method.invoke(Method.java:498),     water.api.Handler.handle(Handler.java:63),     water.api.RequestServer.serve(RequestServer.java:451),     water.api.RequestServer.doGeneric(RequestServer.java:296),     water.api.RequestServer.doGet(RequestServer.java:220),     javax.servlet.http.HttpServlet.service(HttpServlet.java:687),     javax.servlet.http.HttpServlet.service(HttpServlet.java:790),     org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812),     org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587),     org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127),     org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515),     org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061),     org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141),     org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110),     org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97),     water.JettyHTTPD$LoginHandler.handle(JettyHTTPD.java:198),     org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110),     org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97),     org.eclipse.jetty.server.Server.handle(Server.java:499),     org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311),     org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258),     org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544),     org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635),     org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555),     java.lang.Thread.run(Thread.java:748)];parms={limit=-1, src=gs://gcp-public-data-nexrad-l2/2018/01/01/}
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: java.lang.NullPointerException
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.persist.PersistManager.calcTypeaheadMatches(PersistManager.java:251)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.api.TypeaheadHandler.files(TypeaheadHandler.java:12)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at sun.reflect.GeneratedMethodAccessor6.invoke(Unknown Source)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at java.lang.reflect.Method.invoke(Method.java:498)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.api.Handler.handle(Handler.java:63)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.api.RequestServer.serve(RequestServer.java:451)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.api.RequestServer.doGeneric(RequestServer.java:296)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.api.RequestServer.doGet(RequestServer.java:220)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at water.JettyHTTPD$LoginHandler.handle(JettyHTTPD.java:198)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.Server.handle(Server.java:499)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:258)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
04-16 13:49:55.223 10.0.0.9:54321        12199  #31937-22 ERRR: 	at java.lang.Thread.run(Thread.java:748)
04-16 13:49:55.224 10.0.0.9:54321        12199  #31937-22 ERRR: Caught exception: 
04-16 13:49:55.224 10.0.0.9:54321        12199  #31937-22 ERRR: 
04-16 13:49:55.224 10.0.0.9:54321        12199  #31937-22 ERRR: ERROR MESSAGE:
04-16 13:49:55.224 10.0.0.9:54321        12199  #31937-22 ERRR: 
04-16 13:49:55.224 10.0.0.9:54321        12199  #31937-22 ERRR: Caught exception: java.lang.NullPointerException from: water.persist.PersistManager.calcTypeaheadMatches(PersistManager.java:251)
```

Pscheidl commented:
Shows whole stacktrace when a file on GCS in not available with current permissions or is not publicly available for anonymous users. Acceptable. No private information in the stacktrace - acceptable.
